frontend/app/components/routing/wp-list/wp-list.controller.ts

```diff
@@ -98,55 +105,48 @@ function WorkPackagesListController($scope:any,
       }
     });
 
-    wpTableFilters.observeOnScope($scope).subscribe(filters => {
-      updateAndExecuteIfAltered(filters.current, 'filters', true);
-    });
-
-    wpTableGroupBy.observeOnScope($scope).subscribe(groupBy => {
-      updateAndExecuteIfAltered(groupBy.current, 'groupBy', true);
-    });
-
-    wpTableSortBy.observeOnScope($scope).subscribe(sortBy => {
-      updateAndExecuteIfAltered(sortBy.current, 'sortBy', true);
-    });
-
-    wpTableSum.observeOnScope($scope).subscribe(sums => {
-      updateAndExecuteIfAltered(sums.current, 'sums', true);
-    });
-
-    wpTableTimeline.observeOnScope($scope).subscribe(timeline => {
-      updateAndExecuteIfAltered(timeline.current, "timelineVisible");
-    });
-
-    wpTableHierarchies.observeOnScope($scope).subscribe(hierarchies => {
-      updateAndExecuteIfAltered(hierarchies.current, 'showHierarchies');
-    });
-
-    wpTableColumns.observeOnScope($scope).subscribe(columns => {
-      updateAndExecuteIfAltered(columns.current, "columns");
-    });
+    setupChangeObserver(wpTableFilters, 'filters');
+    setupChangeObserver(wpTableGroupBy, 'groupBy');
+    setupChangeObserver(wpTableSortBy, 'sortBy');
+    setupChangeObserver(wpTableSum, 'sums');
+    setupChangeObserver(wpTableTimeline, 'timelineVisible', false);
+    setupChangeObserver(wpTableHierarchies, 'showHierarchies', false);
+    setupChangeObserver(wpTableColumns, 'columns', false);
   }
 
-  function updateAndExecuteIfAltered(updateData:any, name:string, triggerUpdate:boolean = false) {
-    if (isAnyDependentStateClear()) {
-      return;
-    }
-
-    let query = states.table.query.value;
-
-    if (!query || _.isEqual(query[name], updateData)) {
-      return;
-    }
-
-    let pagination = wpTablePagination.current;
-
-    query[name] = _.cloneDeep(updateData);
-
-    states.table.query.putValue(query);
-
-    if (triggerUpdate) {
-      updateResultsVisibly(true);
-    }
+  function setupChangeObserver(service:WorkPackageTableBaseService, name:string, triggerUpdate:boolean = true) {
+    service
+      .observeUntil(scopeDestroyed$($scope))
+      .distinctUntilChanged(
+        (a,b) => _.isEqual(a,b),
+        (stateValue:WorkPackageTableBaseState<any>) => stateValue.extractedCompareValue
+      )
+      .subscribe((stateValue:WorkPackageTableBaseState<any>) => {
+        // Avoid updating while not all states are initialized
+        if (isAnyDependentStateClear()) {
+          return;
+        }
+
+        // Avoid updating if the query isn't yet available
+        const query = states.table.query.value;
+        if (!query) {
+          return;
+        }
+
+        // Avoid updating if the query is up-to-date
+        // (Loaded into query elsewhere)
+        const queryValue = stateValue.comparerFunction()(query[name]);
+        if (_.isEqual(queryValue, stateValue.extractedCompareValue)) {
+          return;
+        }
```

Comment:
Now that you have pushed it this far, how about we place those three conditions in `filter` (maybe each)?